### PR TITLE
Update project to implement RxJava2 natively with RxAndroidBle

### PR DESCRIPTION
**Pull request by shadow7**

### CODE REVIEW

File: library/src/main/kotlin/io/intrepid/bleidiom/BleIdiomImpl.kt [hunk 1]
```diff
@@ -174,10 +178,10 @@ internal class BleCharValueDelegate<Val : Any>() : BleCharHandlerDSL<Val> {
             letMany(uuid, service?.sharedConnection, fromByteArray) { charUUID, connection, transform ->
                 connection.firstOrError().flatMapTry {
                     it.readCharacteristic(charUUID)
-                            .toRx2()
-                            .singleOrError()
-                            .map { byteArray -> transform(byteArray) }
-                            .doOnSuccess { backingField.currentValue = it }
+                        .toObservable()
```
Comment (streetsofboston) [inline library/src/main/kotlin/io/intrepid/bleidiom/BleIdiomImpl.kt]: Not sure, but I think that `readCharacteristic(...)` returns a `Singe`. 
If so, remove `.toObservable().singleOrError()` entirely.

File: library/src/main/kotlin/io/intrepid/bleidiom/BleIdiomImpl.kt [hunk 2]
```diff
@@ -203,26 +207,25 @@ internal class BleCharValueDelegate<Val : Any>() : BleCharHandlerDSL<Val> {
             letMany(uuid, service?.sharedConnection, toByteArray) { charUUID, connection, transform ->
                 connection.firstOrError().flatMapTry {
                     val bytes = transform(value)
-                    val writeObs = if (bytes.size > mtu) {
+                    val writeObs : Observable<ByteArray> = if (bytes.size > mtu) {
```
Comment (streetsofboston) [inline library/src/main/kotlin/io/intrepid/bleidiom/BleIdiomImpl.kt]: This should be a `Single<ByteArray>` and then the `toObservable()` can be removed from line 223.
The `build()` method of the LongWriteBuilder seems to return an Observable, not a Single. This decision by RxAndroidBLE looks odd to me. Still, call `singleOrError` on that instead: `.build().singleOrError()`

File: library/src/main/kotlin/io/intrepid/bleidiom/BleIdiomImpl.kt [hunk 3]
```diff
@@ -203,26 +207,25 @@ internal class BleCharValueDelegate<Val : Any>() : BleCharHandlerDSL<Val> {
...
-                        it.writeCharacteristic(charUUID, bytes)
+                        it.writeCharacteristic(charUUID, bytes).toObservable()
                     }
-                    writeObs.toRx2()
-                            .doOnSubscribe { backingField.inFlightValue = value }
-                            .singleOrError()
-                            .map { _ -> value }
-                            .doOnSuccess { backingField.currentValue = it }
+                    writeObs.doOnSubscribe { backingField.inFlightValue = value }
+                        .singleOrError()
```
Comment (streetsofboston) [inline library/src/main/kotlin/io/intrepid/bleidiom/BleIdiomImpl.kt]: And then the call to `singleOrError()` can be removed.

File: library/src/main/kotlin/io/intrepid/bleidiom/BleIdiomService.kt [hunk 4]
```diff
@@ -181,7 +184,7 @@ open class BleService<Svc : BleService<Svc>> : BleConfigureDSL<Svc> {
     fun observeConnectionState() = device.observeConnectionState()
 
     fun discoverPrimaryServices(timeout: Long = 20_000) =
-            sharedConnection.take(1).flatMapTry { it.discoverServices(timeout, TimeUnit.MILLISECONDS).toRx2() }
+            sharedConnection.take(1).flatMapTry { it.discoverServices(timeout, TimeUnit.MILLISECONDS).toObservable() }
```
Comment (streetsofboston) [inline library/src/main/kotlin/io/intrepid/bleidiom/BleIdiomService.kt]: For some reason, RxAndroidBle folks think `discoverServices` emits a value only once. If this is indeed correct, the return method for `discoverPrimaryServices()` should reflect that. Returning a number (list of values) that completes. Calling `toObservable()` may still be the right thing to do here, though... not sure...

**Review by streetsofboston — CHANGES_REQUESTED** [review]

**Comment by streetsofboston** [comment]
The RxAndroidBle version should be updated to the released version `1.5.0`.
See if the `ConnectionSharingAdapter` can be replaced with the `ReplayingShare` from `com.jakewharton.rx2:replaying-share`.

File: library/src/main/kotlin/io/intrepid/bleidiom/BleIdiomService.kt (hunk 4 above)
Comment (pathakashish) [reply library/src/main/kotlin/io/intrepid/bleidiom/BleIdiomService.kt]: duscoverServices may return multiple values depending on the services device supports. Eg. same device can support temperature monitor and heart rate monitor.
I think toObservable is fine.

File: library/src/main/kotlin/io/intrepid/bleidiom/BleIdiomImpl.kt [hunk 5]
```diff
@@ -205,25 +204,26 @@ internal class BleCharValueDelegate<Val : Any>() : BleCharHandlerDSL<Val> {
         backingField.writeAction = { value ->
             val mtu = service?.mtuSize ?: MIN_MTU_SIZE
             letMany(uuid, service?.sharedConnection, toByteArray) { charUUID, connection, transform ->
-                connection.firstOrError().flatMapTry {
+                connection.firstOrError().flatMapTry { rxBleConnection : RxBleConnection ->
```
Comment (streetsofboston) [inline library/src/main/kotlin/io/intrepid/bleidiom/BleIdiomImpl.kt]: Remove the `: RxBleConnection`. The type can be inferred automatically.

File: library/src/main/kotlin/io/intrepid/bleidiom/BleIdiomImpl.kt [hunk 6]
```diff
@@ -205,25 +204,26 @@ internal class BleCharValueDelegate<Val : Any>() : BleCharHandlerDSL<Val> {
...
                     val bytes = transform(value)
-                    val writeObs : Observable<ByteArray> = if (bytes.size > mtu) {
+                    val writeObs : Single<ByteArray> = if (bytes.size > mtu) {
```
Comment (streetsofboston) [inline library/src/main/kotlin/io/intrepid/bleidiom/BleIdiomImpl.kt]: Remove the `: Single<RxBleConnection>`. The type can be inferred automatically.

File: library/src/main/kotlin/io/intrepid/bleidiom/BleIdiomImpl.kt [hunk 7]
```diff
@@ -205,25 +204,26 @@ internal class BleCharValueDelegate<Val : Any>() : BleCharHandlerDSL<Val> {
...
                         val (batchSize, batchBytes) = toBatchInfo(value, bytes)
                         if (batchSize in 1..mtu) {
-                            it.createNewLongWriteBuilder()
+                            Single.fromObservable<ByteArray> {
```
Comment (streetsofboston) [inline library/src/main/kotlin/io/intrepid/bleidiom/BleIdiomImpl.kt]: Remove the `<ByteArray>`. The type can be inferred automatically.

Or remove the Single.fromObservable entirely and add `singleOrError()` instead:
```
                            rxBleConnection.createNewLongWriteBuilder()
                                    .setCharacteristicUuid(charUUID)
                                    .setMaxBatchSize(batchSize)
                                    .setBytes(batchBytes)
                                    .build()
                                    .singleOrError()
```

File: library/src/main/kotlin/io/intrepid/bleidiom/BleIdiomImpl.kt [hunk 8]
```diff
@@ -205,25 +204,26 @@ internal class BleCharValueDelegate<Val : Any>() : BleCharHandlerDSL<Val> {
...
+                                rxBleConnection.createNewLongWriteBuilder()
                                 .setCharacteristicUuid(charUUID)
                                 .setMaxBatchSize(batchSize)
                                 .setBytes(batchBytes)
                                 .build()
+                            }
                         } else {
-                            Observable.error(IllegalStateException("toBatchInfo must return a " +
+                            Single.error<ByteArray>(IllegalStateException("toBatchInfo must return a " +
```
Comment (streetsofboston) [inline library/src/main/kotlin/io/intrepid/bleidiom/BleIdiomImpl.kt]: Remove the `<ByteArray>`. The type can be inferred automatically.

**Review by streetsofboston — CHANGES_REQUESTED** [review]

File: library/src/main/kotlin/io/intrepid/bleidiom/BleIdiomImpl.kt [hunk 9]
```diff
@@ -190,39 +193,39 @@ internal class BleCharValueDelegate<Val : Any>() : BleCharHandlerDSL<Val> {
...
         }
 
         backingField.writeAction = { value ->
             val mtu = service?.mtuSize ?: MIN_MTU_SIZE
             letMany(uuid, service?.sharedConnection, toByteArray) { charUUID, connection, transform ->
-                connection.firstOrError().flatMapTry {
+                connection.firstOrError().flatMapTry { rxBleConnection : RxBleConnection ->
                     val bytes = transform(value)
-                    val writeObs = if (bytes.size > mtu) {
+                    val writeObs : Single<ByteArray> = if (bytes.size > mtu) {
```
Comment (streetsofboston) [inline library/src/main/kotlin/io/intrepid/bleidiom/BleIdiomImpl.kt]: Remove the `: Single<RxBleConnection>`. The type can be inferred automatically.

File: library/src/main/kotlin/io/intrepid/bleidiom/BleIdiomImpl.kt [hunk 10]
```diff
@@ -190,39 +193,39 @@ internal class BleCharValueDelegate<Val : Any>() : BleCharHandlerDSL<Val> {
...
                         val (batchSize, batchBytes) = toBatchInfo(value, bytes)
                         if (batchSize in 1..mtu) {
-                            it.createNewLongWriteBuilder()
-                                    .setCharacteristicUuid(charUUID)
-                                    .setMaxBatchSize(batchSize)
-                                    .setBytes(batchBytes)
-                                    .build()
+                            Single.fromObservable<ByteArray> {
```
Comment (streetsofboston) [inline library/src/main/kotlin/io/intrepid/bleidiom/BleIdiomImpl.kt]: Remove the `<ByteArray>`. The type can be inferred automatically.

Or remove the `Single.fromObservable` entirely and add `singleOrError()` instead:
```
                            rxBleConnection.createNewLongWriteBuilder()
                                    .setCharacteristicUuid(charUUID)
                                    .setMaxBatchSize(batchSize)
                                    .setBytes(batchBytes)
                                    .build()
                                    .singleOrError()
```

File: library/src/main/kotlin/io/intrepid/bleidiom/BleIdiomImpl.kt [hunk 11]
```diff
@@ -190,39 +193,39 @@ internal class BleCharValueDelegate<Val : Any>() : BleCharHandlerDSL<Val> {
...
+                                rxBleConnection.createNewLongWriteBuilder()
+                                .setCharacteristicUuid(charUUID)
+                                .setMaxBatchSize(batchSize)
+                                .setBytes(batchBytes)
+                                .build()
+                            }
                         } else {
-                            rx.Observable.error(IllegalStateException("toBatchInfo must return a " +
-                                    "batch-size between 1 and $mtu, but is $batchSize instead."))
+                            Single.error<ByteArray>(IllegalStateException("toBatchInfo must return a " +
```
Comment (streetsofboston) [inline library/src/main/kotlin/io/intrepid/bleidiom/BleIdiomImpl.kt]: Remove the `<ByteArray>`. The type can be inferred automatically.

File: library/src/main/kotlin/io/intrepid/bleidiom/test/TestModules.kt [hunk 12]
```diff
@@ -161,19 +182,20 @@ private class TestLongWriteBuilder(private val connection: RxBleConnection) : Rx
...
                                 .writeCharacteristic(char, bytes.sliceArray(start until end))
+                                .toObservable()
                                 .concatMap { Observable.just(idx) }
                     }
                     .map { idx -> idx > 0 }
                     .compose(writeOperationAckStrategy)
                     .repeatWhen { it.takeWhile { numberOfBatchesLeft.decrementAndGet() > 0 } }
-                    .toCompletable()
-                    .andThen(Observable.just<ByteArray>(bytes))
+                    .flatMap { Observable.just<ByteArray>(bytes) }
```
Comment (streetsofboston) [inline library/src/main/kotlin/io/intrepid/bleidiom/test/TestModules.kt]: Remove the `<ByteArray>`. The type can be inferred automatically.

**Review by streetsofboston — CHANGES_REQUESTED** [review]

**Comment by shadow7** [comment]
Looks like adding Jake's ReplayingShare Transformer causes "Already connected to bluetooth device" errors in the BleEndToEnd tests.

**Comment by streetsofboston** [comment]
Looks good. 
Be sure to squash the necessary commits before merging.

**Pull request merged**